File: app/src/main/java/com/github/quentin7b/sn/widget/StickyWidgetRemoteViewsFactory.java
package com.github.quentin7b.sn.widget;

import android.content.Context;
import android.content.Intent;
import android.support.v4.content.ContextCompat;
import android.util.Log;
import android.widget.RemoteViews;
import android.widget.RemoteViewsService;

import com.github.quentin7b.sn.ColorHelper;
import com.github.quentin7b.sn.R;
import com.github.quentin7b.sn.database.DatabaseHelper;
import com.github.quentin7b.sn.database.model.StickyNotification;
import com.github.quentin7b.sn.ui.MainActivity;
import com.j256.ormlite.dao.Dao;

import java.sql.SQLException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class StickyWidgetRemoteViewsFactory implements RemoteViewsService.RemoteViewsFactory {

    private Context context;
    private DatabaseHelper.StickyDao database;
    private List<StickyNotification> notificationList;

    /**
     * Create a widget factory and return it
     *
     * @param context the app context
     */
    public StickyWidgetRemoteViewsFactory(Context context) {
        this.context = context;
        database = new DatabaseHelper(context).getDatabase();
    }

    private void updateWidgetListView() {
        notificationList = database.getAll();
        Collections.sort(notificationList);
    }

    @Override
    public void onCreate() {
        updateWidgetListView();
    }

    @Override
    public void onDataSetChanged() {
        updateWidgetListView();
    }

    @Override
    public void onDestroy() {
        notificationList.clear();
    }

    @Override
    public int getCount() {
        return (notificationList != null)
                ? notificationList.size()
                : 0;
    }

    @Override
    public RemoteViews getViewAt(int position) {
        RemoteViews remoteView = new RemoteViews(context.getPackageName(), R.layout.widget_row);
        StickyNotification notification = notificationList.get(position);
        remoteView.setTextViewText(R.id.note_title_tv, notification.getTitle());
        remoteView.setTextColor(R.id.note_title_tv, ContextCompat.getColor(context, ColorHelper.getDefconColor(notification.getDefcon())));
        remoteView.setTextViewText(R.id.note_description_tv, notification.getContent());

        if (!notification.isNotification()) {
            remoteView.setImageViewBitmap(R.id.note_notif_iv, null);
        }

        Intent fillInIntent = new Intent()
                .setAction(MainActivity.ACTION_NOTIFICATION)
                .putExtra(MainActivity.EXTRA_NOTIFICATION, notification);
        remoteView.setOnClickFillInIntent(R.id.widget_row, fillInIntent);

        return remoteView;
    }

    @Override
    public RemoteViews getLoadingView() {
        // Auto
        return null;
    }

    @Override
    public int getViewTypeCount() {
        return 1;
    }

    @Override
    public long getItemId(int position) {
        return notificationList.get(position).getId();
    }

    @Override
    public boolean hasStableIds() {
        return true;
    }
}

File: app/src/main/kotlin/com/github/quentin7b/sn/widget/StickyWidgetRemoteViewsFactory.kt
package com.github.quentin7b.sn.widget

import android.content.Context
import android.content.Intent
import android.opengl.Visibility
import android.os.Bundle
import android.support.v4.content.ContextCompat
import android.util.Log
import android.view.View
import android.widget.RemoteViews
import android.widget.RemoteViewsService

import com.github.quentin7b.sn.ColorHelper
import com.github.quentin7b.sn.R
import com.github.quentin7b.sn.database.DatabaseHelper
import com.github.quentin7b.sn.database.model.StickyNotification
import com.github.quentin7b.sn.ui.MainActivity
import com.j256.ormlite.dao.Dao

import java.sql.SQLException
import java.util.ArrayList
import java.util.Collections

class StickyWidgetRemoteViewsFactory(private val context: Context) : RemoteViewsService.RemoteViewsFactory {

    private val database: DatabaseHelper.StickyDao = DatabaseHelper(context).database
    private var notificationList: ArrayList<StickyNotification> = ArrayList()

    override fun onCreate() {
        // nothing to do
    }

    override fun onDataSetChanged() {
        notificationList.clear()
        notificationList.addAll(database.all.sorted())
    }

    override fun onDestroy() {
        notificationList.clear()
    }

    override fun getCount(): Int {
        return notificationList.size
    }

    override fun getViewAt(position: Int): RemoteViews {
        val remoteView = RemoteViews(context.packageName, R.layout.widget_row)
        val notification = notificationList[position]
        remoteView.setTextViewText(R.id.note_title_tv, notification.title)
        remoteView.setTextColor(R.id.note_title_tv, ContextCompat.getColor(context, ColorHelper.getDefconColor(notification.defcon)))
        remoteView.setTextViewText(R.id.note_description_tv, notification.content)

        remoteView.setViewVisibility(R.id.note_notif_iv,
                if (notification.isNotification)
                    View.VISIBLE
                else
                    View.INVISIBLE
        )

        Log.i("Widget", "Put extra $notification for $position")
        val fillInIntent = Intent()
                .putExtra(MainActivity.EXTRA_NOTIFICATION_ID, notification.id)
        remoteView.setOnClickFillInIntent(R.id.widget_item, fillInIntent)

        return remoteView
    }

    override fun getLoadingView(): RemoteViews? {
        return null
    }

    override fun getViewTypeCount(): Int {
        return 1
    }

    override fun getItemId(position: Int): Long {
        return notificationList[position].id.toLong()
    }

    override fun hasStableIds(): Boolean {
        return true
    }
}
